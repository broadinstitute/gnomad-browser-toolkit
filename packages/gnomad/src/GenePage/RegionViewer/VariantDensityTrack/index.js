/* eslint-disable space-before-function-paren */
/* eslint-disable no-shadow */
/* eslint-disable comma-dangle */
/* eslint-disable import/no-unresolved */
/* eslint-disable import/extensions */
/* eslint-disable no-case-declarations */
/* eslint-disable react/prop-types */

import React, { PropTypes } from 'react'
import R from 'ramda'
import { getTableIndexByPosition } from 'lens-utilities/lib/variant'
import { range, max } from 'd3-array'
import { line } from 'd3-shape'
import { scaleLinear } from 'd3-scale'

import { connect } from 'react-redux'

import {
  searchFilteredVariants,
} from 'lens-redux-gene-page/lib/resources/table'


import css from './styles.css'

console.log(css)

const VariantAxis = ({ title, height, leftPanelWidth, trackYScale }) => {
  const YTicks = trackYScale ? () => {
    return (
        <g>
          {trackYScale.ticks(3).map(t => {
            return (
              <g key={t}>
                <line
                  x1={leftPanelWidth - 10}
                  x2={leftPanelWidth - 5}
                  y1={trackYScale(t)}
                  y2={trackYScale(t)}
                  stroke={'black'}
                />
                <text
                  x={leftPanelWidth - 30}
                  y={trackYScale(t) + 5}
                >
                  {t}
                </text>
              </g>
            )
          })}
        </g>
    )
  } : null
  const YAxis = () => {
    return (
      <svg width={leftPanelWidth} height={height}>
        <text
          className={css.yLabel}
          x={5}
          y={height / 2}

        >
          {title}
        </text>
        {trackYScale && <YTicks />}
      </svg>
    )
  }
  return (
    <div
      style={{ width: leftPanelWidth }}
    >
      <YAxis/>
      {/*<div className={css.variantAxisName} style={{ fontSize: 12 }}>
        {title}
      </div>*/}
    </div>
  )
}
VariantAxis.propTypes = {
  title: PropTypes.string.isRequired,
  leftPanelWidth: PropTypes.number.isRequired,
}

const VariantDensityTrack = ({
  height = 80,
  width,
  invertOffset,
  variants,
  leftPanelWidth
}) => {
  const makeVariantPlotData = (variants) => {

    const slidingWindowBp = 500

    const variantDensity = range(30, width, 1).map(i => {
      const pos = invertOffset(i)
      const left = getTableIndexByPosition(pos - slidingWindowBp, variants)
      const right = getTableIndexByPosition(pos + slidingWindowBp, variants)
      return { pos, x: i, density: variants.slice(left, right).length / slidingWindowBp }
    })

    const yMax = max(R.pluck('density', variantDensity))
    const densityYScale = scaleLinear()
      .domain([0, yMax])
      .range([height - 25, 10])
    const variantDensityLine = line()
      .defined((base) => {
        return !isNaN(base.density)
      })
      .x(base => base.x)
      .y(base => densityYScale(base.density))

    return {
      variantDensity,
      variantDensityLine,
      densityYScale,
    }
  }

  const jsVariants = variants.toJS()

  const { variantDensityLine, densityYScale, variantDensity } = makeVariantPlotData(jsVariants)

  const renderLine = (data) => {
    return (
      <path
        d={variantDensityLine(data)}
        fill={'none'}
        stroke={'black'}
        opacity={1}
        strokeWidth={1}
      />
    )
  }

  return (
    <div className={css.track}>
      <VariantAxis
        height={height}
        leftPanelWidth={leftPanelWidth}
        title={''}
        trackYScale={densityYScale}
      />
      <svg
        width={width}
        height={height}
      >
      {renderLine(variantDensity)}
    </svg>
  </div>
  )
}

export default connect(state => ({
  variants: searchFilteredVariants(state),
}))(VariantDensityTrack)
