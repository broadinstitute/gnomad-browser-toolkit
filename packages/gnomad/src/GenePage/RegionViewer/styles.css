@import "../../config.css";

:export {
  exonColor: var(--exonColor);
  paddingColor: var(--paddingColor);
  masterExonThickness: 20px;
  masterPaddingThickness: 3px;
  flipOutExonThickness: 25px;
  flipOutPaddingThickness: 5px;
}

.regionViewer {
  display: flex;
  flex-direction: column;
  border: 0 solid #000;
  align-items: center;
  padding-left: 0;
  padding-bottom: 10px;
}

.regionArea {
  flex-direction: column;
  /*border: 1px solid var(--secondaryColor);*/
}

.track {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 0;
}

.transcriptContainer {
  display: flex;
  flex-direction: row;
  padding-bottom: 3px;
  padding-top: 3px;
}

.transcriptLeftAxis {
  display: flex;
  width: 100%;
}

.transcriptName {
  display: flex;
  width: 100%;
}

.transcriptTrackBackground {
  fill: var(--backgroundColor);
}

.transcriptTrackLine {
  stroke: var(--paddingColor);
  /*stroke-width: var(--paddingThickness);*/
}

.transcriptData {
  display: flex;
  justify-content: center;
}

.transcriptsGrouped {
  display: flex;
  flex-direction: column;
}

.transcriptFlipOutButtonContainer {
  display: flex;
  align-items: center;
  height: 100%;
  width: 100%;
  /*border: 1px blue #000;*/
}

.transcriptFlipOutButton {
  font-size: 22px;
  color: var(--primaryColor);
  background-color: var(--backgroundColor);
  border-radius: 5%;
  border: 0;
}

.transcriptFlipOutButton:hover {
  font-size: 22px;
  color: var(--backgroundColor);
  background-color: var(--primaryColor);
  border-radius: 5%;
  border: 0;
}

.coverageTrack {
  display: flex;
}

.coverageYAxis {
  display: flex;
  justify-content: flex-end;
}

.coverageArea {

}

.ylabel {
  font-size: 12px;
  text-anchor: middle;
  fill: var(--primaryColor);
}

.yticktext {
  font-size: 8px;
  text-anchor: end;
  fill: var(--primaryColor);
}

