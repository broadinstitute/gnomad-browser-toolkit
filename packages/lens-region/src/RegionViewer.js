import React, { PropTypes, Component } from 'react'
import injectTapEventPlugin from 'react-tap-event-plugin'

import {
  calculateOffsetRegions,
  calculatePositionOffset,
  invertPositionOffset,
  calculateXScale,
} from '@lens/utils-genomic-coordinates'  // eslint-disable-line

import defaultStyles from './styles.css'

injectTapEventPlugin()

const {
  exonColor,
  paddingColor,
  masterExonThickness,
  masterPaddingThickness,
} = defaultStyles

class RegionViewer extends Component {

  static propTypes = {
    css: PropTypes.object,
    regions: PropTypes.array.isRequired,
    regionAttributes: PropTypes.object,
    padding: PropTypes.number.isRequired,
    exonSubset: PropTypes.array,
    onRegionClick: PropTypes.func,
  }

  static defaultProps = {
    css: defaultStyles,
    exonSubset: null,
    onRegionClick: () => {},
    regionAttributes: {
      CDS: {
        color: exonColor,
        thickness: masterExonThickness,
      },
      start_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      end_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      intron: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      default: {
        color: 'grey',
        thickness: masterPaddingThickness,
      },
    },
  }

  state = {
    leftPanelWidth: 100,
    rightPanelWidth: 50,
    featuresToDisplay: ['CDS'],
    ready: false,
  }

  setWidth = (event, newValue) => {
    const newWidth = 800 * newValue
    this.setState({ width: newWidth })
  }

  setLeftPanelWidth = (event, newValue) => {
    const leftPanelWidth = Math.floor(400 * newValue)
    this.setState({ leftPanelWidth })
  }

  renderChildren = (childProps) => {
    return React.Children.map(this.props.children, (child) => {
      if (child) {
        return React.cloneElement(child, childProps)
      }
    })
  }

  render() {
    const { css } = this.props
    const { featuresToDisplay, leftPanelWidth } = this.state
    const {
      regions,
      regionAttributes,
      width,
      exonSubset,
      padding,
    } = this.props

    const offsetRegions = calculateOffsetRegions(
      featuresToDisplay,
      regionAttributes,
      padding,
      regions,
      exonSubset,
    )

    const positionOffset = calculatePositionOffset(offsetRegions)
    const xScale = calculateXScale(width, offsetRegions)
    const invertOffset = invertPositionOffset(offsetRegions, xScale)

    const childProps = {
      leftPanelWidth,
      positionOffset,
      invertOffset,
      xScale,
      width,
      offsetRegions,
      regionAttributes,
      padding,
    }

    return (
      <div className={css.regionViewer}>
        {/*<p>Exon padding {padding.toPrecision(3)} bp</p>*/}
        <div style={{ width: width + leftPanelWidth }} className={css.regionArea}>
          {this.renderChildren(childProps)}
        </div>
      </div>
    )
  }
}
export default RegionViewer
