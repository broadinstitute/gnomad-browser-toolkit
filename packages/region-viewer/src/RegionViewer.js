import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styled from 'styled-components'

import {
  calculateOffsetRegions,
  calculatePositionOffset,
  invertPositionOffset,
  calculateXScale,
} from './coordinates'

const RegionViewerWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: ${props => props.width}px;
  margin: 0 auto 10px;
  font-size: 12px;
`

const exonColor = '#212121'
const paddingColor = '#BDBDBD'
const masterExonThickness = '25px'
const masterPaddingThickness = '5px'

const defaultRegionAttributes = {
  CDS: {
    color: exonColor,
    thickness: masterExonThickness,
  },
  start_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  end_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  intron: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  default: {
    color: 'grey',
    thickness: masterPaddingThickness,
  },
}

export class RegionViewer extends Component {
  static propTypes = {
    children: PropTypes.node,
    featuresToDisplay: PropTypes.arrayOf(PropTypes.string),
    leftPanelWidth: PropTypes.number,
    padding: PropTypes.number.isRequired,
    regionAttributes: PropTypes.objectOf(
      PropTypes.shape({
        color: PropTypes.string.isRequired,
        thickness: PropTypes.string.isRequired,
      })
    ),
    regions: PropTypes.arrayOf(
      PropTypes.shape({
        feature_type: PropTypes.string.isRequired,
        start: PropTypes.number.isRequired,
        stop: PropTypes.number.isRequired,
      })
    ).isRequired,
    rightPanelWidth: PropTypes.number,
    width: PropTypes.number.isRequired,
  }

  static defaultProps = {
    children: undefined,
    featuresToDisplay: ['CDS'],
    leftPanelWidth: 100,
    regionAttributes: defaultRegionAttributes,
    rightPanelWidth: 160,
  }

  renderChildren(childProps) {
    const { children } = this.props
    return React.Children.map(
      children,
      child => (child ? React.cloneElement(child, childProps) : null)
    )
  }

  render() {
    const {
      featuresToDisplay,
      regions,
      regionAttributes,
      width,
      padding,
      leftPanelWidth,
      rightPanelWidth,
    } = this.props

    const offsetRegions = calculateOffsetRegions(
      featuresToDisplay,
      regionAttributes,
      padding,
      regions
    )

    const positionOffset = calculatePositionOffset(offsetRegions)
    const xScale = calculateXScale(width, offsetRegions)
    const invertOffset = invertPositionOffset(offsetRegions, xScale)

    const childProps = {
      leftPanelWidth,
      positionOffset,
      invertOffset,
      xScale,
      width,
      offsetRegions,
      rightPanelWidth,
    }

    return (
      <RegionViewerWrapper width={width + leftPanelWidth + rightPanelWidth}>
        {this.renderChildren(childProps)}
      </RegionViewerWrapper>
    )
  }
}
