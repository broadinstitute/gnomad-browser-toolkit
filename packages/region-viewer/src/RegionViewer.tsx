import React from 'react'
import styled from 'styled-components'

import { mergeOverlappingRegions, regionViewerScale, Region, ScalePosition } from './coordinates'

export type RegionViewerContextValue = {
  centerPanelWidth: number
  isPositionDefined: (pos: number) => boolean
  leftPanelWidth: number
  regions: Region[]
  rightPanelWidth: number
  scalePosition: ScalePosition
}

const defaultRegionViewerContextValue: RegionViewerContextValue = {
  centerPanelWidth: 1440,
  isPositionDefined: () => false,
  leftPanelWidth: 100,
  regions: [],
  rightPanelWidth: 160,
  scalePosition: regionViewerScale([], [0, 1440])
}

export const RegionViewerContext = React.createContext<RegionViewerContextValue>(
  defaultRegionViewerContextValue
)

type Props = {
  children?: React.ReactNode
  leftPanelWidth: number
  regions: Region[]
  rightPanelWidth: number
  width: number
}

const RegionViewerWrapper = styled.div`
  display: flex;
  flex-direction: column;
`

export const RegionViewer = ({
  children,
  leftPanelWidth = 100,
  regions,
  rightPanelWidth = 160,
  width
}: Props) => {
  const mergedRegions = mergeOverlappingRegions([...regions].sort((r1, r2) => r1.start - r2.start))

  const isPositionDefined = (pos: number) =>
    mergedRegions.some(region => region.start <= pos && pos <= region.stop)

  const centerPanelWidth = width - (leftPanelWidth + rightPanelWidth)
  const scalePosition = regionViewerScale(mergedRegions, [0, centerPanelWidth])

  const childProps = {
    centerPanelWidth,
    isPositionDefined,
    leftPanelWidth,
    regions,
    rightPanelWidth,
    scalePosition
  }

  return (
    <RegionViewerWrapper style={{ width }}>
      <RegionViewerContext.Provider value={childProps}>{children}</RegionViewerContext.Provider>
    </RegionViewerWrapper>
  )
}
