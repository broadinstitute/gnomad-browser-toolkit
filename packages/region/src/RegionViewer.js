import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import {
  calculateOffsetRegions,
  calculatePositionOffset,
  invertPositionOffset,
  calculateXScale,
} from '@broad/utilities/src/coordinates'  // eslint-disable-line

const RegionViewerWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-left: 10px;
  padding-bottom: 10px;
`

const RegionArea = styled.div`
  flex-direction: column;
`

const exonColor = '#212121'
const paddingColor = '#BDBDBD'
const masterExonThickness = '25px'
const masterPaddingThickness = '5px'

class RegionViewer extends Component {
  static propTypes = {
    regions: PropTypes.array.isRequired,
    regionAttributes: PropTypes.object,
    padding: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    leftPanelWidth: PropTypes.number.isRequired,
    exonSubset: PropTypes.array,
    broadcast: PropTypes.func,
    featuresToDisplay: PropTypes.array,
  }

  static defaultProps = {
    exonSubset: null,
    leftPanelWidth: 100,
    onRegionClick: () => {},
    broadcast: () => {},
    featuresToDisplay: ['CDS'],
    regionAttributes: {
      CDS: {
        color: exonColor,
        thickness: masterExonThickness,
      },
      start_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      end_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      intron: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      default: {
        color: 'grey',
        thickness: masterPaddingThickness,
      },
    },
  }

  state = {
    rightPanelWidth: 50,
    ready: false,
  }

  componentWillMount () {
    this.broadcastOffsetRegions()
  }

  componentDidUpdate () {
    this.broadcastOffsetRegions()
  }

  setWidth = (event, newValue) => {
    const newWidth = 800 * newValue
    this.setState({ width: newWidth })
  }

  setLeftPanelWidth = (event, newValue) => {
    const leftPanelWidth = Math.floor(400 * newValue)
    this.setState({ leftPanelWidth })
  }

  broadcastOffsetRegions = () => {
    if (this.props.regions) {
      const offsetRegions = calculateOffsetRegions(
        this.props.featuresToDisplay,
        this.props.regionAttributes,
        this.props.padding,
        this.props.regions,
        this.props.exonSubset
      )
      this.props.broadcast({ offsetRegions })
    }
  }

  renderChildren = (childProps) => {
    // eslint-disable-next-line
    return React.Children.map(this.props.children, (child) => {
      if (child) {
        return React.cloneElement(child, childProps)
      }
    })
  }

  render() {
    const { featuresToDisplay } = this.props
    const {
      regions,
      regionAttributes,
      width,
      exonSubset,
      padding,
      leftPanelWidth,
    } = this.props

    const offsetRegions = calculateOffsetRegions(
      featuresToDisplay,
      regionAttributes,
      padding,
      regions,
      exonSubset
    )

    const positionOffset = calculatePositionOffset(offsetRegions)
    const xScale = calculateXScale(width, offsetRegions)
    const invertOffset = invertPositionOffset(offsetRegions, xScale)

    const childProps = {
      leftPanelWidth,
      positionOffset,
      invertOffset,
      xScale,
      width,
      offsetRegions,
      regionAttributes,
      padding,
    }

    return (
      <RegionViewerWrapper>
        <RegionArea width={width + leftPanelWidth}>
          {this.renderChildren(childProps)}
        </RegionArea>
      </RegionViewerWrapper>
    )
  }
}
export default RegionViewer
