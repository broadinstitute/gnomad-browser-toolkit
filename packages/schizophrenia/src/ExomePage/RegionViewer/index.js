/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable space-before-function-paren */
/* eslint-disable no-shadow */
/* eslint-disable comma-dangle */
/* eslint-disable import/no-unresolved */
/* eslint-disable import/extensions */
/* eslint-disable no-case-declarations */

import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import R from 'ramda'
import { scaleLinear } from 'd3-scale'
import { max } from 'd3-array'

import RegionViewer from 'lens-region'
import TranscriptTrack from 'lens-track-transcript'
import VariantTrack from 'lens-track-variant'
import Navigator from 'lens-redux-gene-page/lib/containers/Navigator'
import { groupExonsByTranscript } from 'lens-utilities/lib/transcriptTools'
import { exonPadding } from 'lens-redux-gene-page/lib/resources/active'
import { geneData } from 'lens-redux-gene-page/lib/resources/genes'
import { searchFilteredVariants as visibleVariants } from 'lens-redux-gene-page/lib/resources/table'

import css from './styles.css'

const {
  exonColor,
  paddingColor,
  masterExonThickness,
  masterPaddingThickness,
} = css

const attributeConfig = {
  CDS: {
    color: '#424242',
    thickness: masterExonThickness,
  },
  start_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  end_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  intron: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  default: {
    color: 'grey',
    thickness: masterPaddingThickness,
  },
}

const factor = 50

const GeneRegion = ({
  gene,
  visibleVariants,
  exonPadding,
}) => {
  const geneJS = gene.toJS()
  const geneExons = geneJS.exons
  const canonicalExons = geneJS.transcript.exons
  const transcriptsGrouped = groupExonsByTranscript(geneExons)

  // function getColor (variant) {
  //
  //   const color = scaleLinear()
  //     .domain([])
  // }

  // const modifiedVariants = visibleVariants.map(v => v.set('color', getColor(v)))
  const modifiedVariants = visibleVariants
  const markerConfigOther = {
    circleRadius: 3,
    circleStroke: 'black',
    circleStrokeWidth: 1,
    yPositionSetting: 'random',
    fillColor: '#757575',

  }

  const lof = ['splice_acceptor_variant', 'splice_donor_variant', 'stop_gained', 'frameshift_variant']
  const missense = ['missense_variant']

  const consequenceCategories = [
    { annotation: 'lof', groups: lof, colour: '#757575' },
    { annotation: 'missense', groups: missense, colour: '#757575' },
  ]

  const factor = 50

  const splitTracks = consequenceCategories.map((consequence, index) => {
    let rowHeight
    const filteredVariants = modifiedVariants.filter(variant =>
      R.contains(variant.transcriptConsequenceTerms, consequence.groups))
    if (filteredVariants.size / factor < 20) {
      rowHeight = 30
    } else {
      rowHeight = filteredVariants.size / factor
    }
    return (
      <VariantTrack
        key={`${consequence.annotation}-${index}`}
        title={`${consequence.annotation} (${filteredVariants.size})`}
        height={rowHeight}
        markerConfig={markerConfigOther}
        variants={filteredVariants}
      />
    )
  })

  const otherVariants = modifiedVariants.filter(v =>
    !R.contains(v.consequence, [...lof, ...missense]))

  let otherHeight
  if (otherVariants.size / factor < 20) {
    otherHeight = 30
  } else {
    otherHeight = otherVariants.size / factor
  }

  const allTrack = (
    <VariantTrack
      key={'All-variants'}
      title={`other (${otherVariants.size})`}
      height={otherHeight}
      color={'#75757'}
      markerConfig={markerConfigOther}
      variants={otherVariants}
    />
  )

  return (
    <div className={css.geneRegion}>
      <RegionViewer
        css={css}
        width={1000}
        padding={exonPadding}
        regions={canonicalExons}
        regionAttributes={attributeConfig}
        leftPanelWidth={100}
      >
        <TranscriptTrack
          css={css}
          transcriptsGrouped={transcriptsGrouped}
          height={10}
        />
        {splitTracks}
        {allTrack}
        {/*<VariantTrack
          key={'odds_ratio'}
          title={''}
          height={100}
          color={'#75757'}
          markerConfig={markerConfigOdds}
          variants={variantsArray}
        />
        <VariantTrack
          key={'scz_af'}
          title={''}
          height={100}
          color={'#75757'}
          markerConfig={markerConfigSczAF}
          variants={variantsArray}
        />
        <VariantTrack
          key={'hc_af'}
          title={''}
          height={100}
          color={'#75757'}
          markerConfig={markerConfigHCAF}
          variants={variantsArray}
        />*/}
        <Navigator noVariants />
      </RegionViewer>
    </div>
  )
}
GeneRegion.propTypes = {
  gene: PropTypes.object.isRequired,
  visibleVariants: PropTypes.any.isRequired,
  exonPadding: PropTypes.number.isRequired,
}
export default connect(state => ({
  gene: geneData(state),
  exonPadding: exonPadding(state),
  visibleVariants: visibleVariants(state),
}))(GeneRegion)
