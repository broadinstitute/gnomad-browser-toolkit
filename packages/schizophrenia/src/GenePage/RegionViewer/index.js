/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable space-before-function-paren */
/* eslint-disable no-shadow */
/* eslint-disable comma-dangle */
/* eslint-disable import/no-unresolved */
/* eslint-disable import/extensions */
/* eslint-disable no-case-declarations */

import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import R from 'ramda'

import RegionViewer from 'lens-region'
import TranscriptTrack from 'lens-track-transcript'
import VariantTrack from 'lens-track-variant'
import Navigator from 'lens-redux-gene-page/lib/containers/Navigator'
import { groupExonsByTranscript } from 'lens-utilities/lib/transcriptTools'
import { exonPadding } from 'lens-redux-gene-page/lib/resources/active'
import { geneData } from 'lens-redux-gene-page/lib/resources/genes'
import { visibleVariants } from 'lens-redux-gene-page/lib/resources/table'

import css from './styles.css'

const {
  exonColor,
  paddingColor,
  masterExonThickness,
  masterPaddingThickness,
} = css

const attributeConfig = {
  CDS: {
    color: '#424242',
    thickness: masterExonThickness,
  },
  start_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  end_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  intron: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  default: {
    color: 'grey',
    thickness: masterPaddingThickness,
  },
}

const factor = 50

const GeneRegion = ({
  gene,
  visibleVariants,
  exonPadding,
}) => {
  const geneJS = gene.toJS()
  const geneExons = geneJS.exons
  const canonicalExons = geneJS.transcript.exons
  const transcriptsGrouped = groupExonsByTranscript(geneExons)

  const markerConfigOther = {
    markerType: 'af',
    circleRadius: 3,
    circleStroke: 'black',
    circleStrokeWidth: 1,
    yPositionSetting: 'random',
    fillColor: '#757575',
    afMax: 0.001,
  }

  const allTrack = (
    <VariantTrack
      key={'All-variants'}
      title={`All variants: (${visibleVariants.length})`}
      height={30}
      color={'#75757'}
      markerConfig={markerConfigOther}
      variants={visibleVariants}
    />
  )

  return (
    <div className={css.geneRegion}>
      <RegionViewer
        css={css}
        width={1150}
        padding={exonPadding}
        regions={canonicalExons}
        regionAttributes={attributeConfig}
      >
        <TranscriptTrack
          css={css}
          transcriptsGrouped={transcriptsGrouped}
          height={10}
        />
        {allTrack}
        <Navigator />
      </RegionViewer>
    </div>
  )
}
GeneRegion.propTypes = {
  gene: PropTypes.object.isRequired,
  visibleVariants: PropTypes.array.isRequired,
  exonPadding: PropTypes.number.isRequired,
}
export default connect(state => ({
  gene: geneData(state),
  exonPadding: exonPadding(state),
  visibleVariants: visibleVariants(state),
}))(GeneRegion)
