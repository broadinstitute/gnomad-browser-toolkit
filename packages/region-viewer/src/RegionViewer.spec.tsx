import React, { useContext } from 'react'
import { expect, test, describe } from '@jest/globals'
import { createRoot } from 'react-dom/client'
import 'jest-styled-components'

import { RegionViewer, RegionViewerContext } from './RegionViewer'

const flushPromises = () => new Promise(resolve => setTimeout(resolve, 0))

const makeFieldConsumer = (fieldName: string) => {
  return () => {
    const context = useContext(RegionViewerContext)
    return <div>{JSON.stringify(context[fieldName])}</div>
  }
}

describe('Track', () => {
  test('makes centerPanelWidth available via context, calculating it by overall width minus panel widths', async () => {
    const Consumer = makeFieldConsumer('centerPanelWidth')
    const container = document.createElement('div')
    const root = createRoot(container)

    root.render(
      <RegionViewer leftPanelWidth={130} rightPanelWidth={320} width={800} regions={[]}>
        <Consumer />
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })

  test("makes isPositionDefined available via context, which indicates if a given position lies within any of the track's regions", async () => {
    const regions = [
      { start: 91, stop: 109 },
      { start: 191, stop: 209 }
    ]

    const Consumer = () => {
      const context = useContext(RegionViewerContext)
      return (
        <>
          <div>{context.isPositionDefined(90) && 'yes'}</div>
          <div>{context.isPositionDefined(100) && 'yes'}</div>
          <div>{context.isPositionDefined(110) && 'yes'}</div>
          <div>{context.isPositionDefined(190) && 'yes'}</div>
          <div>{context.isPositionDefined(200) && 'yes'}</div>
          <div>{context.isPositionDefined(210) && 'yes'}</div>
        </>
      )
    }

    const container = document.createElement('div')
    const root = createRoot(container)
    root.render(
      <RegionViewer leftPanelWidth={130} rightPanelWidth={320} width={800} regions={regions}>
        <Consumer />
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })

  test('makes leftPanelWidth available via context', async () => {
    const Consumer = makeFieldConsumer('leftPanelWidth')
    const container = document.createElement('div')
    const root = createRoot(container)
    root.render(
      <RegionViewer leftPanelWidth={130} rightPanelWidth={320} width={800} regions={[]}>
        <Consumer />
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })

  test('makes regions available via context', async () => {
    const Consumer = makeFieldConsumer('regions')

    const container = document.createElement('div')
    const root = createRoot(container)
    root.render(
      <RegionViewer
        leftPanelWidth={130}
        rightPanelWidth={320}
        width={800}
        regions={[
          { start: 123, stop: 456 },
          { start: 789, stop: 1011 }
        ]}
      >
        <Consumer />
      </RegionViewer>
    )
    expect(container).toMatchSnapshot()
  })

  test('makes rightPanelWidth available via context', async () => {
    const Consumer = makeFieldConsumer('rightPanelWidth')
    const container = document.createElement('div')
    const root = createRoot(container)

    root.render(
      <RegionViewer leftPanelWidth={130} rightPanelWidth={320} width={800} regions={[]}>
        <Consumer />
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })

  test('makes scalePosition available via context', async () => {
    const regions = [
      { start: 91, stop: 109 },
      { start: 191, stop: 209 },
      { start: 321, stop: 381 }
    ]

    const Consumer = () => {
      const context = useContext(RegionViewerContext)
      return (
        <>
          <div>{context.scalePosition(100)}</div>
          <div>{context.scalePosition(110)}</div>
          <div>{context.scalePosition(190)}</div>
          <div>{context.scalePosition(200)}</div>
          <div>{context.scalePosition(210)}</div>
          <div>{context.scalePosition(311)}</div>
          <div>{context.scalePosition(331)}</div>
          <div>{context.scalePosition(382)}</div>
        </>
      )
    }

    const container = document.createElement('div')
    const root = createRoot(container)
    root.render(
      <RegionViewer leftPanelWidth={130} rightPanelWidth={320} width={800} regions={regions}>
        <Consumer />
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })
})
