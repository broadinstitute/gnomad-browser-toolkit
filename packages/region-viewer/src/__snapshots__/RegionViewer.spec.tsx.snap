// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track makes centerPanelWidth available via context, calculating it by overall width minus panel widths 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
    style="width: 800px;"
  >
    <div>
      350
    </div>
  </div>
</div>
`;

exports[`Track makes isPositionDefined available via context, which indicates if a given position lies within any of the track's regions 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
    style="width: 800px;"
  >
    <div />
    <div>
      yes
    </div>
    <div />
    <div />
    <div>
      yes
    </div>
    <div />
  </div>
</div>
`;

exports[`Track makes leftPanelWidth available via context 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
    style="width: 800px;"
  >
    <div>
      130
    </div>
  </div>
</div>
`;

exports[`Track makes regions available via context 1`] = `<div />`;

exports[`Track makes rightPanelWidth available via context 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
    style="width: 800px;"
  >
    <div>
      320
    </div>
  </div>
</div>
`;

exports[`Track makes scalePosition available via context 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
    style="width: 800px;"
  >
    <div>
      31.81818181818182
    </div>
    <div>
      67.17171717171716
    </div>
    <div>
      67.17171717171716
    </div>
    <div>
      98.98989898989899
    </div>
    <div>
      134.34343434343432
    </div>
    <div>
      134.34343434343432
    </div>
    <div>
      169.6969696969697
    </div>
    <div>
      350
    </div>
  </div>
</div>
`;
